Fix registration socket removal on windows

kkmsft opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
3. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change

/kind bug

> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

Fix registration socket removal on windows.

Until the bug - https://github.com/golang/go/issues/33357 is fixed, os.stat wouldn't return the
right mode(socket) on windows. Hence deleting the file, without checking whether
it's a socket, on windows.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

kkmsft commented:
/cc @andyzhangx  @jingxu97 @PatrickLang

msau42 commented:
/ok-to-test
/assign @verult 
who has been reviewing PRs related to registration socket cleanup

ddebroy commented on cmd/csi-node-driver-registrar/node_register.go:
```diff
@@ -36,23 +36,32 @@ func nodeRegister(
-	}
 
 	var oldmask int
 	if runtime.GOOS == "linux" {
```
Instead of the `if` here, how about wrapping up the logic in a function along the lines of `cleanupSocketFile()` and placing that in https://github.com/kubernetes-csi/node-driver-registrar/blob/master/cmd/csi-node-driver-registrar/util_linux.go and https://github.com/kubernetes-csi/node-driver-registrar/blob/master/cmd/csi-node-driver-registrar/util_windows.go ?

kkmsft replied:
makes sense. Will update the PR

kkmsft replied:
Updated the PR. PTAL.

verult replied:
I think Deep's intention was to not have to expose an OS check here. You could have a function where the Linux implementation calls umask() and the Windows implementation does nothing

kkmsft replied:
For the umask will follow up with a short PR removing the checks, to separate out changes to existing code unrelated to this PR's context.

ddebroy commented on cmd/csi-node-driver-registrar/node_register.go:
```diff
@@ -36,23 +36,32 @@ func nodeRegister(
 	var oldmask int
 	if runtime.GOOS == "linux" {
+		fi, err := os.Stat(socketPath)
+		if err == nil && (fi.Mode()&os.ModeSocket) != 0 {
```
Is the `(fi.Mode()&os.ModeSocket) != 0)` useful for Linux? I understand this is exiting code - so q for the maintainers.

ddebroy commented on cmd/csi-node-driver-registrar/node_register.go:
```diff
@@ -36,23 +36,32 @@ func nodeRegister(
+		// right mode(socket) on windows. Hence deleting the file, without checking whether
+		// its a socket, on windows.
+		klog.Warningf("Windows environment detected. Deleting the socket path: %s", socketPath)
+		if err := os.Remove(socketPath); err != nil {
```
You do want to `stat` first for Windows and `os.Remove` only if the path exists, right?

kkmsft replied:
yep, will modify.

kkmsft replied:
Updated the PR. PTAL.

jingxu97 commented on cmd/csi-node-driver-registrar/node_register.go:
```diff
@@ -36,23 +36,32 @@ func nodeRegister(
+		}
 		// Default to only user accessible socket, caller can open up later if desired
 		oldmask, _ = umask(0077)
+	} else {
```
is it better to use 
if runtime.GOOS == "windows" {
...
} else {
...
}

kkmsft replied:
Logic moved to per os util file. so changes in this area. PTAL.

jingxu97 commented on cmd/csi-node-driver-registrar/node_register.go:
```diff
@@ -36,23 +36,32 @@ func nodeRegister(
 		// Default to only user accessible socket, caller can open up later if desired
 		oldmask, _ = umask(0077)
+	} else {
+		// Until the bug - https://github.com/golang/go/issues/33357 is fixed, os.stat wouldn't return the
```
maybe put TODO: at the beginning for easy search later?

kkmsft replied:
Added the TODO in the util_windows.go file. PTAL.

ddebroy commented:
Thanks KK!
/lgtm

kkmsft commented:
@verult - Please take a look.

verult commented on pkg/util/util_linux.go:
```diff
@@ -16,12 +16,29 @@ See the License for the specific language governing permissions and
+
+func CleanupSocketFile(socketPath string) error {
+	fi, err := os.Stat(socketPath)
+	if err == nil && (fi.Mode()&os.ModeSocket) != 0 {
```
FYI this might be buggy: https://github.com/kubernetes-csi/node-driver-registrar/pull/61#discussion_r377326338

We should return an error if the file isn't a socket. But yeah can be a separate PR

verult replied:
@Madhu-1 FYI. IMO this can be reused in the deletion goroutine. We don't *need* the socket check there but it doesn't hurt. If another file type is there, the socket is gone anyway.

kkmsft replied:
Thank you, @verult . Will raise a shorter PR with that change right after this one gets merged.

kkmsft replied:
For the umask will follow up with a short PR removing the checks, since that is existing logic being modified.

verult commented on pkg/util/util_test.go:
```diff
@@ -0,0 +1,110 @@
+	socketPath := "reg.sock"
+	fileExists := false
+
+	// Test1: file does not exist. So clean up should be successful.
```
Could you split each test case into its own function? Easier to see which has failed

kkmsft replied:
Changed to separate tests.

verult commented on pkg/util/util_test.go:
```diff
@@ -0,0 +1,110 @@
+		t.Fatalf("cleanup returned error: %+v", err)
+	}
+
+	// Test2: director does not exist. So clean up should be successful.
```
nit: `director` -> `directory`

kkmsft replied:
Fixed.

kkmsft commented:
@verult, @ddebroy  - Addressed review comments, please take a final look. The lgtm labels got removed since I updated the PR.

I will follow up with two small PR:
1. Fixing the logic in the linux cleanup to return error if a non-socket file is found.
2. Remove the umask GOOS check in the code paths.

@msau42 - Addressed @verult and @ddebroy comments towards lgtm. Addressed comment raised by @jingxu97.  Please take a look and consider for merging.

jingxu97 commented on pkg/util/util_windows.go:
```diff
@@ -0,0 +1,47 @@
+// +build windows
+
+/*
+Copyright 2019 The Kubernetes Authors.
```
2020?

kkmsft replied:
Thank you, @jingxu97 - Updated to 2020.

msau42 commented on pkg/util/util_test.go:
```diff
@@ -0,0 +1,129 @@
+	"k8s.io/klog"
+)
+
+var socketPath = "reg.sock"
```
Generally we create a unique tmp directory per testcase and create test files inside there, so that we can run multiple unit tests in parallel.

kkmsft replied:
@msau42 : Updated the PR to have tmp test directory. PTAL.

msau42 commented on pkg/util/util_test.go:
```diff
@@ -17,51 +17,64 @@ limitations under the License.
-var socketPath = "reg.sock"
+var socketFileName = "reg.sock"
+
+func getRandomTestDir() string {
```
How about using this helper? https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/csi_plugin_test.go#L45

kkmsft replied:
Thank you, @msau42 . Used the helper function and rebased the PR. PTAL.

kkmsft commented:
@msau42 - have fixed the tests, go mod vendor etc. PTAL.

msau42 commented:
Thank you!
/lgtm
/approve
